Added an AutomaticPojoCodecProvider

rozza opened the pull request:
Automatically creates a PojoCodec for any unseen class that contains
serializable / deserializable properties.

JAVA-2567

jyemin commented on bson/src/main/org/bson/codecs/pojo/AutomaticPojoCodec.java:
```diff
@@ -0,0 +1,69 @@
+
+import static java.lang.String.format;
+
+final class AutomaticPojoCodec<T> extends PojoCodec<T> {
```
This class is not referenced anywhere.  What's the intended purpose?

jyemin replied:
And if it turns out not to be necessary than the change to the PojoCodec (to make it abstract) can also be rolled back.

rozza replied:
Argh, I messed up when testing - its needed to wrap the PojoCodec provided by AutomaticCodecProvider - so to provide meaningful error messages if encoding or decoding fails.

rozza replied:
Fixed.

jyemin commented on bson/src/main/org/bson/codecs/pojo/AutomaticPojoCodecProvider.java:
```diff
@@ -0,0 +1,47 @@
+                return null;
+            }
+            pojoCodecProvider = PojoCodecProvider.builder(pojoCodecProvider).register(classModel).build();
+            return PojoCodecProvider.builder(pojoCodecProvider).register(classModel).build().get(clazz, registry);
```
Why not just

    return pojoCodecoProvider.get(class, registry);

?

rozza replied:
Fixed.

jyemin commented on bson/src/main/org/bson/codecs/pojo/AutomaticPojoCodecProvider.java:
```diff
@@ -0,0 +1,47 @@
+ * @since 3.5
+ */
+public final class AutomaticPojoCodecProvider implements CodecProvider {
+    private PojoCodecProvider pojoCodecProvider = PojoCodecProvider.builder().build();
```
The way this field is used is not thread-safe.  It requires synchronization

rozza replied:
Fixed - we no longer need a mutable `PojoCodecProvider`

rozza commented:
Updated and refactored some more.

I was using a static CodecRegistry in the tests and that hid a bug with discriminator lookups and AutomaticPojoCodecs.  I fixed by making the `DiscriminatorLookup` mutable when using AutomaticPojoCodecs. I also noticed I could simplify the `PojoCodec` implementations by removing the need for a `PojoCodecProvider` they now just rely on the registry.

I added an `IllegalStateException` check, so it doesn't throw when it fails to create a `ClassModel` but rather returns null.

**Error scenarios now are:**

* Has no serializable / deserializable properties.
   The provider returns null, so the CodecRegistry will throw an error if no other providers can help.
* Can't create a ClassModel
  Eg the user provides `Document.class` - which is considered an illegal Map implementation with no type information. The provider returns null, so the CodecRegistry will throw an error if no other providers can help.
* Can't encode / decode - the `AutomaticPojoCodec` wraps the exception with a user friendly message, indicating a custom codec may be needed.

I'm pretty happy with the results and think that if a user does something unexpected - they will find out in testing.  They should have an error message indicating either they are missing a codec or they will get an encoding/decoding error indicating that the `AutomaticPojoCodec` failed and they may need to customise the codecs manually.

rozza commented:
Updated based on fridays conversation.  Automatic is now just a setting and no longer in the default registry.

jyemin commented on bson/src/main/org/bson/codecs/pojo/PojoCodecProvider.java:
```diff
@@ -72,6 +74,17 @@ public static Builder builder() {
+            } catch (IllegalStateException e) {
+                return null;
+            }
+            if (!clazz.isInterface() && classModel.getPropertyModels().isEmpty()) {
```
What's the purpose of the interface check?

jyemin replied:
What's the purpose of the interface check?

rozza replied:
Because interfaces can have no properties, we sometimes need to have a class model for them. They do rely on the discriminator to work to rehydrate, which the PojoCodec will handle.

rozza replied:
The example is in the `ConcreteAndNestedAbstractInterfaceModel` class, it has a field that is of type `InterfaceBasedModel` - which we want to be able to handle.

jyemin commented on bson/src/main/org/bson/codecs/pojo/PojoCodecProvider.java:
```diff
@@ -72,6 +74,17 @@ public static Builder builder() {
+            if (!clazz.isInterface() && classModel.getPropertyModels().isEmpty()) {
+                return null;
+            }
+            discriminatorLookup.addClassModel(classModel);
```
I'm not clear why discriminator lookup is handled differently for automatic pojo creation and lazy creation based on a package name match?  Could the two paths be unified?

rozza replied:
Good catch - done.

jyemin commented on bson/src/main/org/bson/codecs/pojo/PojoCodecProvider.java:
```diff
@@ -100,7 +114,21 @@ public PojoCodecProvider build() {
+        }
+
+        /**
+         * Sets whether the provider should automatically try to create a {@link ClassModel} for unknown any classes.
```
"of any class for which it is requested"?

rozza replied:
Updated.
